metaseq/data/streaming_src_tgt_dataset.py

```diff
@@ -0,0 +1,166 @@
+
+    Args:
+        dataset (~torch.utils.data.IterableDataset): dataset to chunk
+        block_size (int): maximum block size
+        break_mode (str, optional): Mode used for breaking tokens. Values can
+            be one of:
+            - 'none': break tokens into equally sized blocks (up to block_size)
+        drop_last (bool, optional): drop the last item (default: False)
+        padding_idx (int, optional): index to use for padding symbols
+            (required if *drop_last* is ``False``)
+        shuffle_buffer_size (int, optional): buffer this many items and shuffle
+            using the provided *seed*; default value is 1, so no shuffling is
+            performed. This can be adjusted dynamically after initialization,
+            but only before iteration has begun.
+        seed (int, optional): seed for shuffling
+    """
+
+    def __init__(
+        self,
+        dataset: torch.utils.data.IterableDataset,
+        block_size: int,
+        break_mode: str = "none",
+        drop_last: Optional[bool] = False,
+        padding_idx: Optional[int] = None,
+        shuffle_buffer_size: int = 1,
+        seed: Optional[int] = None,
+    ):
+        super().__init__()
+        self.dataset = dataset
+        self.block_size = block_size
+        self.break_mode = break_mode
+        self.drop_last = drop_last
+        self.padding_idx = padding_idx
+        self.shuffle_buffer_size = shuffle_buffer_size
+        self.seed = seed
+        if break_mode == "none" or break_mode == "complete":
+            self.block_iterator = yield_src_tgt_blocks
+        elif break_mode == "eos":
+            raise NotImplementedError(f"EOS break mode is currently not implemented!")
+        else:
+            raise NotImplementedError(f"Unknown break mode: {break_mode}")
+
+        if not drop_last and padding_idx is None:
+            raise ValueError("padding_idx is required when drop_last is False")
+
+        assert shuffle_buffer_size >= 1
+        if shuffle_buffer_size > 1 and seed is None:
+            raise ValueError("seed is required when shuffle_buffer_size > 1")
+
+        self._started_iteration = False
+
+    def set_epoch(self, epoch):
+        if hasattr(self.dataset, "set_epoch"):
+            self.dataset.set_epoch(epoch)
+
+    def set_shuffle_buffer_size(self, new_shuffle_buffer_size):
+        assert not self._started_iteration
+        self.shuffle_buffer_size = new_shuffle_buffer_size
+
+    def __iter__(self):
+        assert not self._started_iteration
+        self.started_iteration = True
+
+        block_itr = self.block_iterator(
+            self.dataset,
+            self.block_size,
+            self.drop_last,
+            self.padding_idx,
+        )
+
+        if self.seed is not None:
+            # add a random offset (2273) to the given seed to decouple this RNG
+            # from any other RNG instances elsewhere
+            rng = np.random.default_rng(2273 + self.seed)
```

Comment:
This looks to be the same RNG offset copied from StreamingTokenBlockDataset. Is that intentional?

Reply:
Ah it just needs to be something distinct for decoupling.  Can you pick another random offset to be safe? Opened https://github.com/facebookresearch/metaseq/issues/65 to tackle this more broadly.

Reply:
Changing to self.rng. We dont need to change the number since this dataset is used independently of token_block_dataset.